geometry/parse-ply.cpp

```diff
@@ -0,0 +1,316 @@
+/**
+ * Copyright (c) 2021 Melown Technologies SE
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * *  Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ *
+ * *  Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+/**
+ * @file geometry/parse-ply.cpp
+ * @author Tomas Novak <tomas.novak@melowntech.com>
+ *
+ * PLY parser that allows loading ply files using simple interface
+ * 
+ * Specify your own `PlyParserBase` child and load any ply you want!
+ */
+
+#include "parse-ply.hpp"
+
+#include <boost/algorithm/string.hpp>
+
+namespace geometry
+{
+
+namespace ba = boost::algorithm;
+
+/// Header parsing functions
+
+/// Helper that throws and exception if EOF bit is set
+inline void throwOnEof(const std::ifstream& f)
+{
+    if (f.eof())
+    {
+        LOGTHROW(err4, std::runtime_error)
+            << "Unexpected EOF reached while parsing PLY header";
+    }
+}
+
+/// Helper retrieving the next line of file, throws if EOF reached
+inline void nextLine(std::ifstream& f, std::string& line)
+{
+    std::getline(f, line);
+    ba::trim(line);
+    throwOnEof(f);
+}
+
+/// Splits property line to name and type
+bool parseProperty(const std::string& line, std::vector<PlyProperty>& props)
+{
+    if (!ba::starts_with(line, "property")) { return false; }
+
+    std::size_t a = line.find_first_of(' ');
+    std::size_t b = line.find_last_of(' ');
+    if ((a == std::string::npos) || (a == b) || (b == line.size() - 1))
```

Comment:
The type information from the header could be used for parsing. Also, this check is not entirely correct as it accepts invalid property definitions. See http://paulbourke.net/dataformats/ply/

Reply:
I leave dealing with types for the `PlyParserBase` object - the types are passed along with property names here:
https://github.com/melowntech/libgeometry/blob/d5a46b57597d56779049c8d41899f4b7e152a845/geometry/parse-ply.cpp#L213